### PR TITLE
Add support for dcd_sof_enable() to some additional ports.

### PR DESCRIPTION
**Pull request by andrewleech**
**Describe the PR**
This PR fills out the stubbed dcd_sof_enable() function for mimxrt, samd and renesas-ra ports.

**Additional context**
I'm working on a TinyUSB feature in micropython that uses SOF to provide some simple timing / delay shortly after connection of CDC: https://github.com/micropython/micropython/pull/14485

This also relates to https://github.com/hathach/tinyusb/discussions/2595 where it was found a delay is needed after CDC connection before sending / flushing TX data from TinyUSB (device) to host so its not lost.

In the micropython stm port (not using tinyusb) the SOF interrupt is enabled on CDC  connection / DTR enable and is used to count 8ms before disabling the SOF interrupt again and flushing buffered data to the host. I'm planning on replicating this process using TinyUSB on other micropython ports, eventually replacing the STM USB stack with TinyUSB on that port also.

### CODE REVIEW

File: src/portable/microchip/samd/dcd_samd.c [hunk 1]
```diff
@@ -183,9 +183,12 @@ void dcd_connect(uint8_t rhport)
...
 {
   (void) rhport;
-  (void) en;
 
-  // TODO implement later
+  if (en) {
+    USB->DEVICE.INTENSET.bit.SOF = 1;
+  } else {
+    USB->DEVICE.INTENCLR.bit.SOF = 1; 
+  }
```
Comment (andrewleech) [inline src/portable/microchip/samd/dcd_samd.c:191]: SAMD support has been tested (both enable and disable) on a [Seeed Studio XIAO SAMD21](https://wiki.seeedstudio.com/Seeeduino-XIAO-TinyUSB/)

File: src/portable/chipidea/ci_hs/dcd_ci_hs.c [hunk 2]
```diff
@@ -309,10 +309,12 @@ void dcd_disconnect(uint8_t rhport)
...
-  (void) rhport;
-  (void) en;
-
-  // TODO implement later
+  ci_hs_regs_t* dcd_reg = CI_HS_REG(rhport);
+  if (en) {
+      dcd_reg->USBINTR |= INTR_SOF;
+  } else {
+      dcd_reg->USBINTR &= ~INTR_SOF;
+  }
```
Comment (andrewleech) [inline src/portable/chipidea/ci_hs/dcd_ci_hs.c:317]: This implementation has been tested  (both enable and disable) on NXP MIMXRT-1010-EVK where the interrupt fires at 125ms as expected for a high speed connection.

File: src/portable/nordic/nrf5x/dcd_nrf5x.c [hunk 3]
```diff
@@ -283,9 +286,13 @@ void dcd_connect(uint8_t rhport) {
...
-  (void) en;
-
-  // TODO implement later
+  if (en) {
+    _dcd.sof_enabled = true;
+    NRF_USBD->INTENSET = USBD_INTENSET_SOF_Msk;
+  } else {
+    _dcd.sof_enabled = false;
+    NRF_USBD->INTENCLR = USBD_INTENCLR_SOF_Msk;
+  }
```
Comment (andrewleech) [inline src/portable/nordic/nrf5x/dcd_nrf5x.c:295]: Tested on nrf52840 dongle (PCA10059).

**Comment by HiFiPhile** [comment]
Thanks for your PR, since your are working on it , could you please also add `dcd_event_sof()` to pass event to upper layer ?
Like:
https://github.com/hathach/tinyusb/blob/ccc7a36043e055ded1f478a979a303e694123187/src/portable/synopsys/dwc2/dcd_dwc2.c#L1161

**Comment by andrewleech** [comment]
> could you please also add dcd_event_sof() to pass event to upper layer ?

Is this different? I noticed they already had something like this? 
https://github.com/hathach/tinyusb/blob/ccc7a36043e055ded1f478a979a303e694123187/src/portable/chipidea/ci_hs/dcd_ci_hs.c#L682


My testing from within micropython is using the `tud_sof_cb()` function which is being called when it's enabled at least.

**Comment by HiFiPhile** [comment]
> Is this different? I noticed they already had something like this?

They are similar, except `dcd_event_sof()` pass the frame number which is useful mostly for audio application.

_I need to fix tud_sof_cb() always receive frame_count 0_

**Comment by HiFiPhile** [comment]
I just fixed the CI issue and frame count in tud_sof_cb().

**Comment by andrewleech** [comment]
I've added `dcd_event_sof()` for the 4 ports now covered in this PR. I think I've got the frame count reading correct but haven't had a chance to test any of these yet. I'll get these tested before calling this PR ready.

I haven't tested renesas changes at all yet as I'm chasing some support to get USB working at all on the renesas evk I've got. 

The 4 ports covered at this point are the main ones I'm planning on adding at this time as these (mimx, samd, nrf, renesas) are the only ones I've got access to where the matching micropython ports directly use libusb, other than rp2 where these functions have already been added. 

I will also look at esp32s3 and esp32s2 are a separate issue, I do want that added too but in expecting some complications from IDF integration. If that's in fact trivial I'll add support for this platform too.

**Comment by andrewleech** [comment]
mimx and samd frame_count are tested working, eg.
image

**Comment by HiFiPhile** [comment]
> I will also look at esp32s3 and esp32s2 are a separate issue, I do want that added too but in expecting some complications from IDF integration. If that's in fact trivial I'll add support for this platform too.

esp32 use dcd_dwc2 which already have sof support, it is little more complicate if you want to try:
``` bash
# 1. Source ESP-IDF
# Windows
%userprofile%\esp\esp-idf\export.bat
# Linux
. $HOME/esp/esp-idf/export.sh
# 2. Enter example directory
cd tinyusb/examples/device/hid_composite_freertos
# 3. Run cmake in project directory specifying the board
cmake -DBOARD=espressif_s3_devkitc -B build -G Ninja .
ninja.exe -C build
```

**Comment by andrewleech** [comment]
I would have assumed esp32s2/3 uses this driver which doesn't have these sof functions configured?
* https://github.com/hathach/tinyusb/blob/ccc7a36043e055ded1f478a979a303e694123187/src/portable/espressif/esp32sx/dcd_esp32sx.c#L254
* https://github.com/hathach/tinyusb/blob/ccc7a36043e055ded1f478a979a303e694123187/src/portable/espressif/esp32sx/dcd_esp32sx.c#L828

**Comment by HiFiPhile** [comment]
dcd_esp32sx is old driver still maintained by espressif, a new driver has been created for all devices with dwc2:
[src/portable/synopsys/dwc2/dcd_dwc2.c](https://github.com/hathach/tinyusb/blob/master/src/portable/synopsys/dwc2/dcd_dwc2.c)

**Comment by andrewleech** [comment]
Thanks @HiFiPhile that's really interesting about the esp32sX driver! With that info I've been able to get a build of micropython working on the S3 that's using our existing tinyusb submodule and this driver directly, ignoring most of the esp32 usb implementation. It just took a little bit of extra support to manage switching from USB serial/jtag mode to USB OTG mode, but my test code for that seems to be working pretty well so far.
Being able to use the same shared tinyunb intergration code across all ports in micropython will be a great advantage, I appreciate your input here!

**Comment by andrewleech** [comment]
Thanks for your SOF frame count fix, I've also now finished adding `dcd_event_sof()` for each of the ports I'm targeting here.

I've been able to test all of them and have confirmed the frame counts are passed through as expected. 

In case it helps, the code in using sof interrupt for plus the test printf to confirm SOF frame count is here: https://github.com/andrewleech/micropython/blob/54e0abf5d15441ad7a2a9d8058e45c6f8e193f68/shared/tinyusb/mp_usbd_cdc.c#L124

File: src/portable/renesas/rusb2/dcd_rusb2.c [hunk 4]
```diff
@@ -756,10 +758,9 @@ void dcd_disconnect(uint8_t rhport)
 
 void dcd_sof_enable(uint8_t rhport, bool en)
 {
-  (void) rhport;
-  (void) en;
-
-  // TODO implement later
+  rusb2_reg_t* rusb = RUSB2_REG(rhport);
+  _dcd.sof_enabled = en;
+  rusb->INTENB0_b.SOFE = en ? 1: 0;
```
Comment (andrewleech) [inline src/portable/renesas/rusb2/dcd_rusb2.c:763]: Renesas updates tested on a EK-RA4W1 which had a usb cable added to the appropriate breakout header pins (the USB peripheral actually isn't connected to a USB port on these eval boards).

**Comment by andrewleech** [comment]
Thanks for your support and reviews @HiFiPhile !

**Pull request merged**